File: haskell/2015/day25a/src/Main.hs
module Main ( main ) where

import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, readFile, stderr)
import Text.Parsec
import Text.Parsec.String
import Control.DeepSeq
import System.Clock

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

file :: Parser (Int, Int)
file = do
       _ <- string "To continue, please consult the code grid in the manual.  Enter the code at row "
       row <- many1 digit
       _ <- string ", column "
       col <- many1 digit
       _ <- string "."
       optional newline >> eof
       return (read row, read col)

generate :: (Int, Int) -> Integer
generate (r, c) 
  | r == 1 && c == 1 = 20151125 :: Integer
  | r > 1 && c == 1  = step (generate (1, r - 1))
  | otherwise        = step (generate (r + 1, c - 1))
  where
    step n = (n * 252533) `mod` 33554393

process :: String -> Integer
process content =
    case parse file "" content of
        Left err -> error (show err)
        Right (row, col) -> generate (row, col)


showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 1000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"

main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            start <- getTime Monotonic
            content <- readFile filename
            let result = process content
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname

File: rust/2015/day25a/src/main.rs
use std::env;
use std::fs;
use std::io;
use std::process;

use regex::Regex;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn generate(row: usize, col: usize) -> i64 {
    let mut n = 20151125;
    let mut r = row;
    let mut c = col;
    while !(r == 1 && c == 1) {
        if r >= 1 && c == 1 {
            c = r - 1;
            r = 1;
        } else {
            r += 1;
            c -= 1;
        }
        n = (n * 252533) % 33554393;
    }
    n
}

fn process_file(filename: &str) -> io::Result<i64> {
    let content = fs::read_to_string(filename)?;
    let re = Regex::new(r"To continue, please consult the code grid in the manual.  Enter the code at row (\d+), column (\d+).").unwrap();
    let caps = re.captures(&content).expect("malformed input");
    let row = caps.get(1).unwrap().as_str().parse::<usize>().unwrap();
    let col = caps.get(2).unwrap().as_str().parse::<usize>().unwrap();
    Ok(generate(row, col))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}
